### DISCUSSION TITLE
About configuring the package using nixos modules

### DISCUSSION THREAD

**Discussion by pierrot-lc** [nix-community/kickstart-nix.nvim#12] [Show and tell]
Hi,
I've just spent the day tweaking my own kickstart-nix.nvim config to add the possibility to select a theme to load based on a `theme` parameter.

I am now able to add the flake as a nixos module in my home-manager configuration and simply specify the way I want to build my neovim package:

```nix
{
  nvim-nix = {
    enable = true;
    theme = "catppuccin";
  };
}
```

I figured out that I could share this here, receive some feedback (I'm still new to this), and if @mrcjkb is interested I could make a PR to add the functionnality to the main `kickstart-nix.nvim` repo.

You can find my neovim flake [here](https://github.com/pierrot-lc/nvim-nix), and my home-manager configuration [here](https://github.com/pierrot-lc/dotfiles).

Have a good day!

**Comment by mrcjkb** [comment]
Hey :wave:

Thanks for sharing!

I want to keep kickstart-nix.nvim simple and unopinionated, so I'd rather not add modules or similar abstractions to the template.
[NixVim](https://github.com/nix-community/nixvim) and [nixCats-nvim](https://github.com/BirdeeHub/nixCats-nvim) are alternate projects that come with NixOS/home-manager modules :smile:

**Reply by pierrot-lc** [reply]
Yes I guess you're right! 😇

**Reply by jaycle** [reply]
@pierrot-lc I've tried looking at your dotfiles and they seem really nice. However, I'm so new at this, I was trying to take a smaller step to get kickstart-nix working with home-manager. Shouldn't that be possible?

I assumed you could follow what's in the README, but replace 
```
environment.systemPackages = with pkgs; [
    nvim-pkg # The default package added by the overlay
];
```

with `home.packages = with pkgs;`. However, it seems this doesn't work. `init.lua` and none of the lua plugins make it to the derivation destination.

If there's something unique to make the overlay correctly patch home-manager, without taking on full HM-style configuration, I'm all ears. 

Forgive me for hijacking an old thread - it just had some context.

**Reply by mrcjkb** [reply]
@jaycle you need to make sure the derivation is built with nixpkgs-unstable (ideally the ref that is pinned in the project's lockfile). Otherwise, the `mkNeovim` function might be broken.
For example, if you import this flake and have something like `inputs.nixpkgs.follows = "nixpkgs"`, it might build with an older version of nixpkgs.

**Reply by pierrot-lc** [reply]
Hi, I added more detailed examples [here](https://github.com/pierrot-lc/nvim-nix). Hopefully it will help you understand what you may have done wrong. Do not hesitate to have a look at [my personal dotfiles](https://github.com/pierrot-lc/dotfiles/tree/main) to see how I use it daily.

If none of this help, you can show us your flake.nix, we should be able to help you out :)

**Reply by jaycle** [reply]
Thank you both. I cleaned up my config and pushed to github.
- [home-manager flake](https://github.com/jaycle/home.nix/blob/master/flake.nix)
- [nvim-pkg](https://github.com/jaycle/home.nix/blob/master/home.nix#L26)

I've got clean/unchanged copy of the template being pointed to [here](https://github.com/jaycle/kickstart-nix.nvim). Though I get the same behavior if I just point directly to this repo.

EDIT: got it working when I changed to
```nix
home.packages = with pkgs; [
  # ...
  # was nvim-pkg
  inputs.kickstart-nvim.packages.${pkgs.system}.default
]  
```

Not sure why the overlay doesn't work :thinking:. I got this idea from [another discussion](https://github.com/nix-community/kickstart-nix.nvim/discussions/16#discussioncomment-9106153)

**Reply by mrcjkb** [reply]
Glad to hear it's working :tada: 

> Not sure why the overlay doesn't work

It's probably because the packages in the overlay are built using whichever nixpkgs you are applying the overlay to. If the `wrapNeovimUnstable` function of that nixpkgs instance doesn't have the signature expected by `mkNeovim.nix`, it won't build correctly.

For instance, older versions of `wrapNeovimUnstable` didn't support passing in a lua RC with `luaRcContent` attribute.

EDIT: I've just pushed a fix that should make sure the overlay uses the correct `wrapNeovimUnstable` as long as the flake is imported with a compatible nixpkgs pin.

**Reply by jaycle** [reply]
Thanks so much for engaging on this. Even though it works using the package directly, I'd still love to get it working the "right" way. As well as close some gaps in my Nix understanding.

I've updated to point to the latest version. 
```nix
    nixpkgs-unstable.url = "nixpkgs/nixos-unstable";
    kickstart-nvim = {
      url = "github:nix-community/kickstart-nix.nvim";
      inputs.nixpkgs.follows = "nixpkgs-unstable";
    };
```

lock section (showing latest github rev):
```json
"kickstart-nvim": {
  "inputs": {
    "flake-utils": "flake-utils",
    "gen-luarc": "gen-luarc",
    "nixpkgs": [
      "nixpkgs-unstable"
    ]
  },
  "locked": {
    "lastModified": 1713974282,
    "narHash": "sha256-QWU5JpIdANDEilnO9qfrHwhxIMPgFn7IrnuaVA8SoP4=",
    "owner": "nix-community",
    "repo": "kickstart-nix.nvim",
    "rev": "52d5cb634521786bd92ac3a79e2b3f724fb57362",
    "type": "github"
  },
  "original": {
    "owner": "nix-community",
    "repo": "kickstart-nix.nvim",
    "type": "github"
  }
},
```

The overlay still doesn't seem to be applying. Does my `nixpkgs` input _need_ to be unstable, instead? Does `inputs.nixpkgs.follows` not work how I think?

If I remove the overlay, then `nvim-pkg` does not exist. And if I remove `nvim-pkg` then I don't have access to the binary. So the inclusion of the overlay _does_ seem to change the availability of neovim. It just doesn't bring in all of the lua/plugins.

By the way, I also moved all this over into my configuration.nix instead, just as a test. I see the same behavior. In both cases, my `nixpkgs` points to 23.11.

**Reply by mrcjkb** [reply]
Darn, I pinned the wrong function :sweat_smile: . Just pushed another fix: https://github.com/nix-community/kickstart-nix.nvim/commit/2662a7caa51d0c1752b0a7bbd3bde934c3abf7f8

> In both cases, my nixpkgs points to 23.11.

This should in theory only be an issue if your `kickstart-nix.nvim` input has `inputs.nixpkgs.follows = "nixpkgs";`.
If you remove that, or you have `inputs.nixpkgs.follows = "nixpkgs-unstable";`, it should in theory work (as long as your nixpkgs-unstable pin has a compatible `wrapNeovimUnstable` function).

**Reply by jaycle** [reply]
You rock! It works! Thanks so much!

**Reply by mrcjkb** [reply]
Thanks for the donation ❤️
